hamilton/function_modifiers/macros.py

```diff
@@ -876,3 +856,322 @@ def optional_config(self) -> Dict[str, Any]:
+        """
+        super(post_pipe, self).__init__()
+        self.transforms = transforms
+        self.collapse = collapse
+        self.chain = _chain
+        self.namespace = namespace
+
+        if self.collapse:
+            raise NotImplementedError(
+                "Collapsing step() functions as one node is not yet implemented for pipe(). Please reach out if you want this feature."
+            )
+
+        if self.chain:
+            raise NotImplementedError("@flow() is not yet supported -- this is ")
+
+    def transform_node(
+        self, node_: node.Node, config: Dict[str, Any], fn: Callable
+    ) -> Collection[node.Node]:
+        """Injects nodes into the graph.
+
+        We create a copy of the original function and rename it to `function_name_raw` to be the
+        initial node. Then we create a node for each step in `post-pipe` and chain them together.
+        The last node is an identity to the previous one with the original name `function_name` to
+        represent an exit point of `post_pipe`.
+        """
+
+        original_node = node_.copy_with(name=f"{node_.name}_raw")
+
+        def __identity(foo: Any) -> Any:
+            return foo
+
+        transforms = self.transforms + (step(__identity).named(fn.__name__),)
+
+        nodes, _ = chain_nodes(
+            _first_arg=original_node.name,
+            _transforms=transforms,
+            _namespace=self.namespace,
+            _config=config,
+            _fn=fn,
+        )
+
+        last_node = nodes[-1].copy_with(name=f"{node_.name}", typ=nodes[-2].type)
+
+        out = [original_node]
+        out.extend(nodes[:-1])
+        out.append(last_node)
+        return out
+
+    def validate(self, fn: Callable):
+        """Validates the the individual steps work together."""
+        for applicable in self.transforms:
+            applicable.validate(
+                chain_first_param=True, allow_custom_namespace=self.namespace is ...
+            )
+        # TODO -- validate that the types match on the chain (this is de-facto done later)
+
+    def optional_config(self) -> Dict[str, Any]:
+        """Declares the optional configuration keys for this decorator.
+        These are configuration keys that can be used by the decorator, but are not required.
+        Along with these we have *defaults*, which we will use to pass to the config.
+
+        :return: The optional configuration keys with defaults. Note that this will return None
+        if we have no idea what they are, which bypasses the configuration filtering we use entirely.
+        This is mainly for the legacy API.
+        """
+        out = {}
+        for applicable in self.transforms:
+            for resolver in applicable.resolvers:
+                out.update(resolver.optional_config)
+        return out
+
+
+# Should this become an abstract class for pipe
```

Comment:
I think it's pretty clean to keep it as a helper function -- abstract classes are good when you have more than one implementation that share state/data. In this case it's one tool that it's built around and it would take some wrangling to get it to work with a class.